src/Graphs/DirectedGraphBuilder.cs

```diff
@@ -0,0 +1,210 @@
+            }
+
+            directSuccessors[from].Add(to);
+            directPredecessors[to].Add(from);
+
+            if (sources.Contains(to))
+            {
+                sources.Remove(to);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Remove(element))
+            {
+                return this;
+            }
+            sources.Remove(element);
+
+            foreach (var to in directSuccessors[element])
+            {
+                directPredecessors[to].Remove(element);
+                if (directPredecessors[to].Count == 0)
+                {
+                    sources.Add(to);
+                }
+            }
+            foreach (var from in directPredecessors[element])
+            {
+                directSuccessors[from].Remove(element);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveArrow(T from, T to)
+        {
+            if (from == null) throw new ArgumentNullException(nameof(from));
+            if (to == null) throw new ArgumentNullException(nameof(to));
+            if (!elements.Contains(from)) throw new ArgumentException("Element not in graph.", nameof(from));
+            if (!elements.Contains(to)) throw new ArgumentException("Element not in graph.", nameof(to));
+
+            if (!directSuccessors[from].Remove(to))
+            {
+                return this;
+            }
+
+            directPredecessors[to].Remove(from);
+            if (directPredecessors[to].Count == 0)
+            {
+                sources.Add(to);
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Creates an immutable directed graph instance.
+        /// </summary>
+        public IDirectedGraph<T> CreateGraph()
+        {
+            return new AdjacencyListDiGraph<T>(
+                ImmutableHashSet.CreateRange(elements),
+                directSuccessors.ToImmutableDictionary(
+                    pair => pair.Key,
+                    pair => ImmutableHashSet.CreateRange(pair.Value)),
+                directPredecessors.ToImmutableDictionary(
+                    pair => pair.Key,
+                    pair => ImmutableHashSet.CreateRange(pair.Value)));
+        }
+
+        /// <summary>
+        /// Creates an immutable directed acyclic graph instance.
+        ///
+        /// <para>It will verify that no cycles exist.</para>
+        /// </summary>
+        public IDirectedAcyclicGraph<T> CreateAcyclicGraph()
+        {
+            if (isGraphCyclic())
+            {
+                throw new InvalidOperationException("Cannot create an directed acyclic graph with cycles.");
+            }
+
+            return CreateAcyclicGraphUnsafe();
+        }
+
+        private bool isGraphCyclic()
```

Comment:
Worth exposing publicly? /shrug

Reply:
I don't think so, but feel free to convince me otherwise.